File: 0743-network-delay-time/0743-network-delay-time.py
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        adj = collections.defaultdict(list)
        for u,v,w in times:
            adj[u].append((v, w))
        
        minhp, vis, t = [(0, k)], set(), 0
        
        while minhp:
            tm, node = heappop(minhp)
            if node in vis:
                continue
            vis.add(node)
            t = tm
            
            for n2, t2 in adj[node]:
                if n2 not in vis:
                    heappush(minhp, (tm+t2, n2))
        
        return t if len(vis)==n else -1

File: 0743-network-delay-time/0743-network-delay-time.java
class Solution {
    public int networkDelayTime(int[][] times, int n, int k) {
        List<int[]>[] adj = new ArrayList[n+1];
        for(int i=0; i<n+1; ++i)
            adj[i] = new ArrayList<>();
        for(int[] t: times){
            adj[t[0]].add(new int[]{t[1], t[2]});
        }
        
        int[] minTime = new int[n+1];
        Arrays.fill(minTime, Integer.MAX_VALUE);
        int maxTime = 0;
        int vis = 0;
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        pq.add(new int[]{0, k});
        
        while(!pq.isEmpty()){
            int currTime = pq.peek()[0];
            int node = pq.peek()[1];
            pq.poll();
            if(minTime[node] < currTime)
                continue;
            maxTime = currTime;
            minTime[node] = currTime;
            vis++;
            
            for(int i=0; i<adj[node].size(); ++i){
                int nbr = adj[node].get(i)[0];
                int nbrTime = currTime + adj[node].get(i)[1];
                if(minTime[nbr]>nbrTime){
                    minTime[nbr] = nbrTime;
                    pq.offer(new int[]{nbrTime, nbr});
                }
            }
        }
        System.out.println(Arrays.toString(minTime));
        return vis>=n ? maxTime : -1;
    }
}